src/webui/api/torrentscontroller.cpp

```diff
@@ -873,15 +873,47 @@ void TorrentsController::editTrackerAction()
 
 void TorrentsController::removeTrackersAction()
 {
-    requireParams({u"hash"_s, u"urls"_s});
+    requireParams({u"urls"_s});
 
-    const auto id = BitTorrent::TorrentID::fromString(params()[u"hash"_s]);
-    BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->getTorrent(id);
-    if (!torrent)
-        throw APIError(APIErrorType::NotFound);
+    const QString hashParam = params()[u"hash"_s];
+    const QStringList urlsParam = params()[u"urls"_s].split(u'|', Qt::SkipEmptyParts);
 
-    const QStringList urls = params()[u"urls"_s].split(u'|');
-    torrent->removeTrackers(urls);
+    // remove trackers from specified torrent
+    if (!hashParam.isEmpty())
+    {
+        const auto id = BitTorrent::TorrentID::fromString(hashParam);
+        BitTorrent::Torrent *const torrent = BitTorrent::Session::instance()->getTorrent(id);
+        if (!torrent)
+            throw APIError(APIErrorType::NotFound);
+
+        torrent->removeTrackers(urlsParam);
+    }
+    // remove trackers from all torrents
+    else
+    {
+        QStringList urls;
+        urls.reserve(urlsParam.size());
+        for (const QString &urlStr : urlsParam)
+            urls << QUrl::fromPercentEncoding(urlStr.toUtf8());
```

Comment:
I doubt using `toUtf8()` is correct. Shouldn't it be `toLatin1()` instead?

Reply:
A valid URL can contain UTF-8 characters.

https://stackoverflow.com/questions/912811/what-is-the-proper-way-to-url-encode-unicode-characters

Reply:
@Piccirello 
If you are going to use `QUrl::fromPercentEncoding` then the input is assumed to be percent encoded and that means no unicode characters exist in it.

Reply:
I also doubt the `urls` parameter can be assumed to be percent encoded.

Reply:
> If you are going to use `QUrl::fromPercentEncoding` then the input is assumed to be percent encoded and that means no unicode characters exist in it.

That's what I mean above.

Reply:
> I also doubt the `urls` parameter can be assumed to be percent encoded.

It looks like you haven't been following the related discussions...
It should be to be percent encoded. Moreover, it should be required by our WebAPI specifications that each URL within list is percent encoded. Otherwise, we cannot connect multiple URLs in one parameter in an unambiguous way.

Reply:
>It looks like you haven't been following the related discussions...

Yeah I didn't.

>It should be to be percent encoded. Moreover, it should be required by our WebAPI specifications that each URL within list is percent encoded. Otherwise, we cannot join multiple URLs in one parameter in an unambiguous way.

A follow up question: shouldn't the API check and reply error if the parameter is not properly percent encoded?

Also I would expect the API documentation on the wiki will mention the requirement of percent encoded.

Reply:
> shouldn't the API check and reply error if the parameter is not properly percent encoded?

I don't think this is achievable in the general case (at least without excessive costs).

Reply:
> > shouldn't the API check and reply error if the parameter is not properly percent encoded?
> 
> I don't think this is achievable in the general case (at least without excessive costs).

Maybe the following code will be suitable for these purposes?
```c++
for (const QString &urlStr : urlsParam)
{
    if (!QUrl(urlStr, QUrl::StrictMode).isValid())
        throw APIError(APIErrorType::BadParams);
}
```

Reply:
> If you are going to use QUrl::fromPercentEncoding then the input is assumed to be percent encoded and that means no unicode characters exist in it.

Good point. Will switch to `toLatin1()`.

Reply:
> Maybe the following code will be suitable for these purposes?
> 
> ```c++
> for (const QString &urlStr : urlsParam)
> {
>     if (!QUrl(urlStr, QUrl::StrictMode).isValid())
>         throw APIError(APIErrorType::BadParams);
> }
> ```

👍 It should do the job.
The question is, should it really be checked and reply back? Should the API have this responsibility?